File: xxf/src/main/java/com/google/android/material/bottomsheet/BottomSheetUtils.java
package com.google.android.material.bottomsheet;

import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.viewpager.widget.ViewPager;
import androidx.viewpager2.widget.ViewPager2;

import com.xxf.arch.R;

/**
 * @version 2.3.1
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @Description
 * @date createTime：2018/9/7
 * 处理 bottomsheet 嵌套viewpager
 */
public final class BottomSheetUtils {

    public static void setupViewPager(ViewPager viewPager) {
        /**
         * 避免重复添加
         */
        Object viewpageListener = viewPager.getTag(R.id.navigation_page_bottom_sheet_listener);
        if (!(viewpageListener instanceof BottomSheetViewPagerListener)) {
            final View bottomSheetParent = findBottomSheetParent(viewPager);
            if (bottomSheetParent != null) {
                viewpageListener = new BottomSheetViewPagerListener(viewPager, bottomSheetParent);
                viewPager.addOnPageChangeListener((ViewPager.OnPageChangeListener) viewpageListener);
                viewPager.setTag(R.id.navigation_page_bottom_sheet_listener, viewpageListener);
            } else if (!viewPager.isAttachedToWindow()) {
                Object attachListener = viewPager.getTag(R.id.navigation_page_bottom_sheet_attach_state_listener);
                if (!(attachListener instanceof View.OnAttachStateChangeListener)) {
                    attachListener = new View.OnAttachStateChangeListener() {
                        @Override
                        public void onViewAttachedToWindow(View v) {
                            setupViewPager((ViewPager) v);
                        }

                        @Override
                        public void onViewDetachedFromWindow(View v) {

                        }
                    };
                    viewPager.addOnAttachStateChangeListener((View.OnAttachStateChangeListener) attachListener);
                    viewPager.setTag(R.id.navigation_page_bottom_sheet_attach_state_listener, attachListener);
                }
            }
        }
    }

    public static void setupViewPager(ViewPager2 viewPager) {
        /**
         * 避免重复添加
         */
        Object viewpageListener = viewPager.getTag(R.id.navigation_page_bottom_sheet_listener);
        if (!(viewpageListener instanceof BottomSheetViewPager2Listener)) {
            final View bottomSheetParent = findBottomSheetParent(viewPager);
            if (bottomSheetParent != null) {
                viewpageListener = new BottomSheetViewPager2Listener(viewPager, bottomSheetParent);
                viewPager.registerOnPageChangeCallback((ViewPager2.OnPageChangeCallback) viewpageListener);
                viewPager.setTag(R.id.navigation_page_bottom_sheet_listener, viewpageListener);
            } else if (!viewPager.isAttachedToWindow()) {
                Object attachListener = viewPager.getTag(R.id.navigation_page_bottom_sheet_attach_state_listener);
                if (!(attachListener instanceof View.OnAttachStateChangeListener)) {
                    attachListener = new View.OnAttachStateChangeListener() {
                        @Override
                        public void onViewAttachedToWindow(View v) {
                            setupViewPager((ViewPager2) v);
                        }

                        @Override
                        public void onViewDetachedFromWindow(View v) {

                        }
                    };
                    viewPager.addOnAttachStateChangeListener((View.OnAttachStateChangeListener) attachListener);
                    viewPager.setTag(R.id.navigation_page_bottom_sheet_attach_state_listener, attachListener);
                }
            }
        }
    }

    private static class BottomSheetViewPagerListener extends ViewPager.SimpleOnPageChangeListener {
        private final ViewPager viewPager;
        private final AndroidBottomSheetBehavior<View> behavior;

        private BottomSheetViewPagerListener(ViewPager viewPager, View bottomSheetParent) {
            this.viewPager = viewPager;
            this.behavior = (AndroidBottomSheetBehavior<View>) AndroidBottomSheetBehavior.from(bottomSheetParent);
        }

        @Override
        public void onPageSelected(int position) {
            viewPager.post(behavior::invalidateScrollingChild);
        }
    }

    private static class BottomSheetViewPager2Listener extends ViewPager2.OnPageChangeCallback {
        private final ViewPager2 viewPager;
        private final AndroidBottomSheetBehavior<View> behavior;

        private BottomSheetViewPager2Listener(ViewPager2 viewPager, View bottomSheetParent) {
            this.viewPager = viewPager;
            this.behavior = (AndroidBottomSheetBehavior<View>) AndroidBottomSheetBehavior.from(bottomSheetParent);
        }

        @Override
        public void onPageSelected(int position) {
            viewPager.post(behavior::invalidateScrollingChild);
        }
    }

    private static View findBottomSheetParent(final View view) {
        View current = view;
        while (current != null) {
            final ViewGroup.LayoutParams params = current.getLayoutParams();
            if (params instanceof CoordinatorLayout.LayoutParams && ((CoordinatorLayout.LayoutParams) params).getBehavior() instanceof AndroidBottomSheetBehavior) {
                return current;
            }
            final ViewParent parent = current.getParent();
            current = parent == null || !(parent instanceof View) ? null : (View) parent;
        }
        return null;
    }

}

File: xxf/src/main/java/androidx/viewpager/widget/BottomSheetUtils.kt
package androidx.viewpager.widget

import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.viewpager.widget.ViewPager.SimpleOnPageChangeListener
import androidx.viewpager2.widget.ViewPager2
import com.xxf.arch.R

/**
 * 用法: 先加layout_behavior
 *     <androidx.viewpager.widget.ViewPager
android:id="@+id/vp"
android:layout_width="match_parent"
android:layout_height="wrap_content"
android:background="@color/grey_db"
app:layout_behavior="androidx.viewpager.widget.ViewPagerBottomSheetBehavior" />


<androidx.viewpager2.widget.ViewPager2
android:id="@+id/vp"
android:layout_width="match_parent"
android:layout_height="wrap_content"
android:background="@color/grey_db"
app:layout_behavior="androidx.viewpager.widget.ViewPagerBottomSheetBehavior" />

 */
object BottomSheetUtils {

    /**
     * 是否是在bottomSheet容器中
     */
    fun isInBottomSheetLayout(view: View?): Boolean {
        var current = view;
        while (current != null) {
            if (current.id == R.id.design_bottom_sheet) {
                return true
            }
            current = current.parent as? View
        }
        return false
    }

    fun setupViewPager(vp: ViewPager) {
        findBottomSheetParent(vp)?.also {
            vp.addOnPageChangeListener(
                BottomSheetViewPagerListener(
                    vp, ViewPagerBottomSheetBehavior.from(it)
                )
            )
        }
    }

    @JvmStatic
    fun getCurrentViewWithVP(vp: ViewPager): View? {
        val currentItem: Int = vp.currentItem
        (0 until vp.childCount).forEach { index ->
            val child = vp.getChildAt(index)
            val layoutParams = child.layoutParams as ViewPager.LayoutParams
            layoutParams.position
            val position = layoutParams.position
            if (layoutParams.isDecor.not() && currentItem == position) {
                return child
            }
        }
        return null
    }

    private fun findBottomSheetParent(view: View): View? {
        var current: View? = view
        while (current != null) {
            val params = current.layoutParams
            if (params is CoordinatorLayout.LayoutParams && params.behavior is ViewPagerBottomSheetBehavior<*>) {
                return current
            }
            val parent = current.parent
            current = if (parent !is View) null else parent
        }
        return null
    }

    private class BottomSheetViewPagerListener(
        private val vp: ViewPager,
        private val behavior: ViewPagerBottomSheetBehavior<View>
    ) : SimpleOnPageChangeListener() {

        override fun onPageSelected(position: Int) {
            vp.post(behavior::invalidateScrollingChild)
        }

    }


    fun setupViewPager(vp: ViewPager2) {
        findBottomSheetParent(vp)?.also {
            vp.registerOnPageChangeCallback(
                BottomSheetViewPagerListener2(
                    vp,
                    ViewPagerBottomSheetBehavior.from(it)
                )
            )
        }
    }

    private class BottomSheetViewPagerListener2(
        private val vp: ViewPager2,
        private val behavior: ViewPagerBottomSheetBehavior<View>
    ) : ViewPager2.OnPageChangeCallback() {

        override fun onPageSelected(position: Int) {
            vp.post(behavior::invalidateScrollingChild)
        }
    }

}